components/butterfly/src/rumor/dat_file.rs

```diff
@@ -46,166 +56,156 @@ pub struct DatFile {
     header:      Header,
     header_size: u64,
     path:        PathBuf,
+    reader:      BufReader<File>,
 }
 
 impl DatFile {
-    pub fn new<T: AsRef<Path>>(member_id: &str, data_path: T) -> Self {
-        DatFile { path:        data_path.as_ref().join(format!("{}.rst", member_id)),
-                  header_size: 0,
-                  header:      Header::default(), }
+    pub fn read_or_create(data_path: PathBuf, server: &Server) -> Result<Self> {
+        let file = OpenOptions::new().create(true)
+                                     .read(true)
+                                     .write(true)
+                                     .open(&data_path)
+                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?;
+        let size = file.metadata()
+                       .map_err(|err| Error::DatFileIO(data_path.clone(), err))?
+                       .len();
+        let reader = BufReader::new(file);
+        let dat_file = DatFile { path: data_path,
+                                 header_size: 0,
+                                 header: Header::default(),
+                                 reader };
+
+        if size == 0 {
+            dat_file.write(server)?;
+        }
+
+        Ok(dat_file)
+    }
+
+    pub fn read(data_path: &Path) -> io::Result<Self> {
+        Ok(DatFile { header:      Default::default(),
+                     header_size: Default::default(),
+                     path:        data_path.to_path_buf(),
+                     reader:      BufReader::new(File::open(&data_path)?), })
     }
 
     pub fn path(&self) -> &Path { &self.path }
 
-    /// # Locking
-    /// * `MemberList::entries` (read) This method must not be called while any MemberList::entries
-    ///   lock is held.
-    pub fn read_into_mlr(&mut self, server: &Server) -> Result<()> {
-        let mut version = [0; 1];
-        let mut size_buf = [0; 8];
-        // JW: Resizing this buffer is terrible for performance, but it's the easiest way to
-        // read exactly N bytes from a file. I'm not sure what the right approach is but this
-        // won't be a performance issue for a long time anyway, if ever.
-        let mut rumor_buf: Vec<u8> = vec![];
-        let mut bytes_read = 0;
-        let file = File::open(&self.path).map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-        let mut reader = BufReader::new(file);
-        reader.read_exact(&mut version)
-              .map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
+    pub fn read_header(&mut self, version: &mut [u8]) -> Result<()> {
```

Comment:
Requiring the caller to know the internal details, like the size of the header seems very error prone. Instead, why not make `read_header` a private method which is always called when a `DatFile` value is created?

Additionally, there's no need for the consumers to know the version. If we change `offset_for_rumor` to return an `Option<64>` rather than panicking when `message_id` is an rumor type, then callers don't have any need for it.

Reply:
Yeah good call